### DISCUSSION TITLE
Build and Run Discussion

### DISCUSSION THREAD

**Discussion by slominskir** [389ds/389-ds-base#6072] [General]
Hi,
Would be nice if the build and run instructions were easier and more portable.  I'm usually on Windows (Ubuntu WSL), Mac, or RHEL.  The build instructions are tailored for Fedora or CentOS.   Ideally the dev environment just runs in a container.   This is already happening with GitHub Action, but it still requires a little work to adapt to run on a local container runtime during development.  For one there are multiple containers, one for compile, then multiple for testing.  Would be nice to consolidate to a single dev container that is not tied to GitHub Action (I guess there might be a run GitHub action locally plugin?).

There is also ["devcontainer"](https://containers.dev/) spec, but that might be unnecessary and too tightly tied to GitHub/Microsoft/VisualStudioCode.

I am actually trying to build/test changes to dscontainer.  I've extended the ci compile container and added some dscontainer runtime pieces in a [dev-container-discussion](https://github.com/slominskir/389-ds-base/tree/dev-container-discussion) branch.  Trying to get a simple process, and this is what I have so far:

```
git clone https://github.com/slominskir/389-ds-base -b dev-container-discussion
cd 389-ds-base
docker compose -f dev.yaml up
```

```
docker exec -it bash dirsrv
cd 389-ds-base
./build.sh
/usr/libexec/dirsrv/dscontainer -r &
```

Currently just bind mounting the source into the container.  Remote dev via ports requires more tooling, but might avoid filesystem permission issues.

Build references:
- https://github.com/389ds/389-ds-base?tab=readme-ov-file#building
https://www.port389.org/docs/389ds/development/building.html
- https://www.port389.org/docs/389ds/contributing.html

Container references:
- https://github.com/389ds/389-ds-base/blob/main/.github/workflows/compile.yml
- https://github.com/389ds/389-ds-base/blob/main/.github/workflows/pytest.yml
- https://github.com/389ds/ci-images
- https://build.opensuse.org/package/view_file/home:firstyear/389-ds-container/Dockerfile?expand=1

Thoughts?

Regards,

Ryan

**Comment by vashirov** [comment]
Hi @slominskir,

Our main target platform is RHEL. So our development/testing workflows are heavily skewed towards RHEL on BM/VM.
GH Actions use containers because there are no Fedora/CentOS runners (see https://github.com/actions/runner-images/issues/2307 for example). So we have to run tests in Fedora containers on Ubuntu VMs. This setup doesn't represent our target platform and has known issues. RH doesn't currently offer RHDS in a container, so there is a very little incentive for us to develop and test in a container environment.

That said, I see how having a devcontainer can make it easier for others to contribute. So feel free to submit your PR for a review. I think it'd be pretty cool to make it work on WSL or Mac.

**Comment by bsehgal** [comment]
How do I build DS-389 latest branch on RHEL9 ?
I really need the latest version for RHEL9 ready for deployment.

Tried the steps in the guide but its failing on ./autogen.sh itself.
./autogen.sh: line 43: shift: shift count out of range
./autogen.sh: line 44: shift: shift count out of range
./autogen.sh: line 48: [: : integer expression expected
./autogen.sh: line 49: [: : integer expression expected
./autogen.sh: line 52: [: : integer expression expected
./autogen.sh: line 53: [: : integer expression expected
./autogen.sh: line 56: [: : integer expression expected
./autogen.sh: line 57: [: : integer expression expected
Found valid libtool version
Running autoreconf -fvi
./autogen.sh: line 103: autoreconf: command not found

And via package upgrade steps its still showing these are latest packages for RHEL9 ?
389-ds-base-libs-2.5.2-8.el9_5.x86_64
389-ds-base-2.5.2-8.el9_5.x86_64

Any hints ?

**Reply by vashirov** [reply]
Looks like you don't have the required build dependencies installed. See this page https://www.port389.org/docs/389ds/contributing.html#building-the-project

> And via package upgrade steps its still showing these are latest packages for RHEL9 ?
389-ds-base-libs-2.5.2-8.el9_5.x86_64
389-ds-base-2.5.2-8.el9_5.x86_64

Yes, that's correct, these are the latest packages available in 9.5 z-stream. Why do you need the latest version?

**Reply by bsehgal** [reply]
There are some deadlocks happening with 2.5.2 version making LDAP stuck. So need the upgraded package looks like its fixed in there.
The link share do not talk much about RHEL9 building process. Can you share the same ? 
Will there be a package published for RHEL9 (2.5.2) was the last one I can download so far.

**Reply by vashirov** [reply]
> There are some deadlocks happening with 2.5.2 version making LDAP stuck. So need the upgraded package looks like its fixed in there.

What issue are you talking about? 389-ds-base-2.5.2-8.el9_5 has fixes for some deadlocks, but I don't know if it the same issue as yours. Please open a separate ticket.
